src/bounded_queue.ml

```diff
@@ -0,0 +1,113 @@
+(*
+  structure node t value: data type, next: pointer to node t
+  structure queue t Head: pointer to node t, Tail: pointer to node t, H lock: lock type, T lock: lock type
+*)
+
+type 'a node = Nil | Next of 'a * 'a node ref
+
+type 'a t = {
+  mutable head : 'a node ref;
+  mutable tail : 'a node ref;
+  head_lock : Mutex.t;
+  tail_lock : Mutex.t;
+  not_empty : Condition.t;
+  not_full : Condition.t;
+  size : int Atomic.t;
+  capacity : int;
```

Comment:
There are use cases for being able to (read and) change the capacity of a bounded queue.  For example, setting the capacity to 0 effectively closes the queue.

Reply:
what would the function signature for such a method look like (val modify :  'a t -> int -> unit)? I assume you would need to hold both locks in this case. also the check for pushing would need to modified to t.size >= t.capacity correct?

Reply:
Yes, something like

```ocaml
val set_capacity : 'a t -> int -> unit
```

And, yes, the logic would then be that when `size < capacity` there is room in the data structure.  Otherwise adding more elements should not be allowed, but removing elements should still work.